broker-daemon/interchain/index.js

```diff
@@ -0,0 +1,275 @@
+ * upstream when an unrecoverable error is encountered.
+ *
+ * @private
+ * @param {string}  hash                  - Base64 string of the hash for the
+ *                                          swap
+ * @param {Payment} inboundPayment        - Expected inbound payment
+ * @param {Engine}  inboundPayment.engine - Engine of the expected inbound
+ *                                          payment
+ * @param {Payment} outboundPayment       - Outbound payment we will make to
+ *                                          retrieve the preimage.
+ * @returns {string}                        Base64 encoded preimage for the swap
+ */
+async function translateOnce (hash, inboundPayment, outboundPayment) {
+  const {
+    paymentPreimage,
+    permanentError
+  } = await getPreimage(hash, inboundPayment, outboundPayment)
+
+  if (permanentError) {
+    logger.error(permanentError)
+    logger.error('Downstream payment encountered an error, cancelling ' +
+      `upstream invoice for ${hash}`)
+    await inboundPayment.engine.cancelSwap(hash)
+
+    const err = new Error(permanentError)
+    err.isPermanent = true
+    throw err
+  }
+
+  logger.debug(`Successfully retrieved preimage for swap ${hash}`)
+
+  logger.debug(`Settling upstream payment for ${hash}`)
+  await inboundPayment.engine.settleSwap(paymentPreimage)
+  logger.debug(`Successfully settled upstream payment for ${hash}`)
+
+  return paymentPreimage
+}
+
+/**
+ * Translate a swap cross-chain and return its preimage, accounting for
+ * temporary errors that may leave us in a non-atomic state.
+ *
+ * @public
+ *
+ * @param {string}  hash                  - Base64 string of the hash for the
+ *                                          swap
+ * @param {Payment} inboundPayment        - Expected inbound payment
+ * @param {Payment} outboundPayment       - Outbound payment we will make to
+ *                                          retrieve the preimage.
+ * @returns {string}                        Base64 encoded preimage for the swap
+ */
+async function translateSwap (hash, inboundPayment, outboundPayment) {
+  try {
+    // we need to await this promise - if we return it, any thrown errors
+    // will be swallowed.
+    const preimage = await translateOnce(hash, inboundPayment, outboundPayment)
+    return preimage
+  } catch (e) {
+    // A permanent error means we are safe to cancel translation and return
+    // the error to the call site.
+    if (e.isPermanent) {
+      logger.error('Permanent Error encountered while translating swap',
+        { error: e.stack, hash })
+      throw e
+    }
+
+    logger.error('Temporary Error encountered while translating swap',
+      { error: e.stack, hash })
+
+    // A temporary error means we don't know the current state, so we need
+    // to restart the whole process
+    logger.debug(`Retrying swap translation for ${hash}`)
+    return translateSwap(hash, inboundPayment, outboundPayment)
```

Comment:
should there be a delay if we continue to retry? I'd assume that there is the potential to get stuck in a loop here

Reply:
Yes there is. Simple delay? Back-off? What do you think?